core/src/autogluon/core/hpo/ray_hpo.py

```diff
@@ -0,0 +1,457 @@
+        To provide custom trial_dir, pass a custom function to `trial_dirname_creator` as kwargs.
+        For example of creating the custom function, refer to `_trial_dirname_creator`.
+    ray_tune_adapter
+        Adapter to provide necessary custom info to ray tune.
+    total_resources
+        Total resources can be used for HPO.
+        If not specified, will use all the resources by default.
+    minimum_cpu_per_trial
+        Specify minimum number of cpu to use per trial.
+    minimum_gpu_per_trial
+        Specify minimum number of gpu to use per trial.
+        Ray supports usage of fraction of gpu.
+    model_estimate_memory_usage
+        Provide optional estimate of the model memory usage.
+        Calculation of the resources_per_trial might use this info to better distribute resources
+    time_budget_s
+        Time limit for the HPO.
+    verbose
+        0 = silent, 1 = only status updates, 2 = status and brief trial results, 3 = status and detailed trial results.
+    **kwargs
+        Additional args being passed to tune.run
+    """
+    assert mode in [MIN, MAX], f'mode is {mode}'
+    if isinstance(hyperparameter_tune_kwargs, str):
+        assert hyperparameter_tune_kwargs in ray_tune_adapter.presets, f'{hyperparameter_tune_kwargs} is not a valid option. Options are {ray_tune_adapter.presets.keys()}'
+        hyperparameter_tune_kwargs = ray_tune_adapter.presets.get(hyperparameter_tune_kwargs)
+    num_samples = hyperparameter_tune_kwargs.get('num_trials', hyperparameter_tune_kwargs.get('num_samples', None))
+    if num_samples is None:
+        num_samples = 1 if time_budget_s is None else 1000  # if both num_samples and time_budget_s are None, we only run 1 trial
+    if not any(isinstance(search_space[hyperparam], (Space, Domain)) for hyperparam in search_space):
+        raise EmptySearchSpace
+    searcher = _get_searcher(hyperparameter_tune_kwargs, metric, mode, supported_searchers=ray_tune_adapter.get_supported_searchers())
+    scheduler = _get_scheduler(hyperparameter_tune_kwargs, supported_schedulers=ray_tune_adapter.get_supported_schedulers())
+    search_space = _convert_search_space(search_space, searcher)
+
+    if ray.is_initialized():
+        # shutdown to reinitialize resources because different model might require different total resources
+        ray.shutdown()
+    ray.init(log_to_driver=False, **total_resources)
+
+    resources_per_trial = hyperparameter_tune_kwargs.get('resources_per_trial', None)
+    if resources_per_trial is None:
+        resources_per_trial = ray_tune_adapter.get_resources_per_trial(
+            total_resources=total_resources,
+            num_samples=num_samples,
+            minimum_gpu_per_trial=minimum_gpu_per_trial,
+            minimum_cpu_per_trial=minimum_cpu_per_trial,
+            model_estimate_memory_usage=model_estimate_memory_usage
+        )
+    else:
+        if isinstance(ray_tune_adapter, AutommRayTuneAdapter):
```

Comment:
The adapter cannot be `None`. The logic here is if `resource_per_trial` is not None, means user wants to provide their own `resource_per_trial`, then we don't calculate it ourselves. Otherwise, we do the calculation. However, if it's `AutommRayTuneAdapter`, we do calculation no matter if user provided `resource_per_trial` or not

Reply:
Good point. I'll pass `resource_per_trial` to each adapter's `get_resources_per_trial()` and let them decide what to do. For example, `TabularAdapter` could simply do nothing. This way we don't need to change the code if there are more adapters in the future